tempesta_fw/apm.c

```diff
@@ -0,0 +1,1039 @@
+ * optimizations.
+ *
+ * Return 0 if new calculation of percentiles is required.
+ * Return 1 if the percentile values don't need the recalculation.
+ */
+static int
+tfw_apm_rblst_update(TfwApmData *data)
+{
+	unsigned long jtmnow = jiffies, jtmwstart, entry_cnt;
+	TfwApmRBuf *rbuf = &data->rbuf;
+	TfwApmRBLst *rblst = &data->rblst;
+	int entry = (jtmnow / tfw_apm_jtmintrvl) % rbuf->rbufsz;
+	TfwApmRBufEnt *crbent = &rbuf->rbent[entry];
+
+	/* The start of the current time window. */
+	jtmwstart = jtmnow - (jtmnow % tfw_apm_jtmintrvl) - tfw_apm_jtmwindow;
+
+	/*
+	 * If the latest percentiles are for a different time window,
+	 * then a recalculation is in order.
+	 */
+	if (unlikely(rblst->jtmwstamp != jtmwstart))
+		return tfw_apm_rblst_buildnew(data, entry, jtmwstart);
+
+	/* The latest percentiles are for the current time window.
+	 * In some cases a recalculation is not required. In some
+	 * other cases the recalculation set up can be simpler.
+	 */
+
+	/* Nothing to do if the current entry is outdated. */
+	if (crbent->jtmistamp < jtmwstart)
+		return 1;
+
+	/* Recalculate if the current entry's data is new. */
+	if (rblst->entry_cnt == 0)
+		return tfw_apm_rblst_buildnew(data, entry, jtmwstart);
+
+	/* Nothing to do if there were no stats updates. */
+	entry_cnt = atomic64_read(&crbent->pcntrng.total_cnt);
+	if (rblst->entry_cnt == entry_cnt)
+		return 1;
+	BUG_ON(rblst->entry_cnt > entry_cnt);
+
+	/* Update the current entry's ranges data. */
+	BUG_ON(rblst->rblstent[0].pcntrng != &crbent->pcntrng);
+	tfw_apm_rngctl_copy(&rblst->rblstent[0], crbent);
+
+	/* Update the counts incrementally. */
+	rblst->total_cnt += entry_cnt - rblst->entry_cnt;
+	rblst->entry_cnt = entry_cnt;
+
+	return 0;
+}
+
+/*
+ * Calculate the latest percentiles if necessary.
+ *
+ * Return 0 if potentially new percentile values were calculated.
+ * Return 1 if the percentile values didn't need the recalculation.
+ */
+static int
+__tfw_apm_calc(TfwApmData *data, TfwApmPrcntl *prcntl)
+{
+	TfwApmRBLst *rblst = &data->rblst;
+	int i, rblstsz = rblst->rblstsz;
+
+	if (tfw_apm_rblst_update(data))
+		return 1;
+
+	/* Clear the percentiles if no updates for the whole time window. */
+	if (unlikely(!rblst->rblstsz)) {
+		if (!rblstsz)
+			return 1;
+		for (i = 0; i < prcntl->prcntlsz; ++i)
+			prcntl->percentile[i].val = 0;
+		return 0;
+	}
+
+	tfw_apm_prnctl_calc(rblst, prcntl);
+
+	return 0;
+}
+
+/*
+ * Calculate the latest percentiles if necessary.
+ *
+ * Return 0 if potentially new percentile values were calculated.
+ * Return 1 if the percentile values didn't need recalculation.
+ * Set the flag to indicate if potentially new values were calculated.
+ */
+static int
+tfw_apm_calc(TfwApmData *data)
+{
+	int ret, wridx;
+	Percentile percentile[] = {
+		{tfw_apm_prcntl_ith[0]}, {tfw_apm_prcntl_ith[1]},
+		{tfw_apm_prcntl_ith[2]}, {tfw_apm_prcntl_ith[3]},
+		{tfw_apm_prcntl_ith[4]}, {tfw_apm_prcntl_ith[5]},
```

Comment:
It'd be better to use `memcpy()` - the initialization looks ugly and is unsupportable.